In [1]:
import pandas as pd
import numpy as np
import missingno as mno
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline

#ignore warning messages 
import warnings
warnings.filterwarnings('ignore')

In [2]:
location=("C:\\Users\\krishna meghana\\Downloads\\diabetes.csv")
data = pd.read_csv(location)

In [3]:
data.head()

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0            6      148             72             35        0  33.6   
1            1       85             66             29        0  26.6   
2            8      183             64              0        0  23.3   
3            1       89             66             23       94  28.1   
4            0      137             40             35      168  43.1   

   DiabetesPedigreeFunction  Age  Outcome  
0                     0.627   50        1  
1                     0.351   31        0  
2                     0.672   32        1  
3                     0.167   21        0  
4                     2.288   33        1

In [4]:
print("The shape of the data is ",data.shape)

The shape of the data is  (768, 9)


In [5]:
#Replacing zeroes with Nan value

data_copy = data.copy(deep = True)
data_copy[['Glucose','BloodPressure','SkinThickness','Insulin','BMI']] = data_copy[['Glucose','BloodPressure','SkinThickness','Insulin','BMI']].replace(0,np.NaN)

## showing the count of Nans

print(data_copy.isnull().sum())

Pregnancies                   0
Glucose                       5
BloodPressure                35
SkinThickness               227
Insulin                     374
BMI                          11
DiabetesPedigreeFunction      0
Age                           0
Outcome                       0
dtype: int64


In [6]:
#For easy access, renaming data_copy as df
df=data_copy

# Highly skewed
df['BMI'].fillna(df['BMI'].median(), inplace = True)
df['SkinThickness'].fillna(df['SkinThickness'].median(), inplace = True)
df['Insulin'].fillna(df['Insulin'].median(), inplace = True)

#Normal
df['Glucose'].fillna(df['Glucose'].mean(), inplace = True)
df['BloodPressure'].fillna(df['BloodPressure'].mean(), inplace = True)

In [45]:
print(df.isnull().sum())

Pregnancies                 0
Glucose                     0
BloodPressure               0
SkinThickness               0
Insulin                     0
BMI                         0
DiabetesPedigreeFunction    0
Age                         0
Outcome                     0
dtype: int64


####  No missing values

In [7]:
#Splitting the dataset into X and y
X=df.drop(columns = ['Outcome'])
y=df['Outcome']

### SVD

In [9]:
X_array=X.to_numpy()

In [10]:
X_array

array([[  6.   , 148.   ,  72.   , ...,  33.6  ,   0.627,  50.   ],
       [  1.   ,  85.   ,  66.   , ...,  26.6  ,   0.351,  31.   ],
       [  8.   , 183.   ,  64.   , ...,  23.3  ,   0.672,  32.   ],
       ...,
       [  5.   , 121.   ,  72.   , ...,  26.2  ,   0.245,  30.   ],
       [  1.   , 126.   ,  60.   , ...,  30.1  ,   0.349,  47.   ],
       [  1.   ,  93.   ,  70.   , ...,  30.4  ,   0.315,  23.   ]])

In [11]:
from numpy import array
from scipy.linalg import svd

In [12]:
U,S,VT=svd(X_array)

In [13]:
U

array([[-3.50515994e-02,  2.89267287e-02,  1.92515525e-02, ...,
        -3.18478515e-02, -1.57014171e-02, -6.68856477e-03],
       [-2.82443900e-02,  1.35534577e-03, -3.58582612e-02, ...,
        -1.07213660e-02, -4.50565098e-02, -3.57634848e-02],
       [-3.70252537e-02,  3.58986709e-02,  9.09422679e-02, ...,
        -6.70992335e-04, -1.24822595e-02,  3.63041762e-02],
       ...,
       [-3.01115788e-02,  2.08123418e-02,  5.59295519e-03, ...,
         9.97060283e-01,  4.38541802e-04, -8.84755492e-04],
       [-3.21326052e-02,  1.67273032e-02,  1.42690281e-02, ...,
         9.25301392e-04,  9.90807954e-01, -3.18730299e-04],
       [-2.91179271e-02,  4.87658295e-03, -2.98378700e-02, ...,
        -6.90417201e-04, -3.87445009e-04,  9.96715641e-01]])

In [14]:
U.shape

(768, 768)

In [15]:
S

array([6069.04932641, 1556.98463775,  489.51781155,  298.27555262,
        246.83993131,  138.9937805 ,   77.49533482,    8.99626975])

In [16]:
S_diag=np.diag(S)
S_diag

array([[6069.04932641,    0.        ,    0.        ,    0.        ,
           0.        ,    0.        ,    0.        ,    0.        ],
       [   0.        , 1556.98463775,    0.        ,    0.        ,
           0.        ,    0.        ,    0.        ,    0.        ],
       [   0.        ,    0.        ,  489.51781155,    0.        ,
           0.        ,    0.        ,    0.        ,    0.        ],
       [   0.        ,    0.        ,    0.        ,  298.27555262,
           0.        ,    0.        ,    0.        ,    0.        ],
       [   0.        ,    0.        ,    0.        ,    0.        ,
         246.83993131,    0.        ,    0.        ,    0.        ],
       [   0.        ,    0.        ,    0.        ,    0.        ,
           0.        ,  138.9937805 ,    0.        ,    0.        ],
       [   0.        ,    0.        ,    0.        ,    0.        ,
           0.        ,    0.        ,   77.49533482,    0.        ],
       [   0.        ,    0.        ,    

In [17]:
S_diag.shape

(8, 8)

In [18]:
VT

array([[-1.69436643e-02, -5.52287652e-01, -3.14482558e-01,
        -1.28148554e-01, -7.33270085e-01, -1.42256677e-01,
        -2.10415948e-03, -1.46436071e-01],
       [ 2.25412521e-02,  5.58593566e-01,  3.90642256e-01,
         1.38252656e-01, -6.78438590e-01,  1.55399657e-01,
         1.87381329e-03,  1.76970240e-01],
       [-3.74755380e-02,  6.17296396e-01, -6.71633484e-01,
        -2.25182153e-01, -4.27613781e-02, -2.26090776e-01,
        -1.17531957e-03, -2.50589803e-01],
       [ 1.48169942e-01, -1.47966410e-02, -1.51885075e-01,
        -2.81784563e-01,  6.48458823e-03, -3.11947157e-01,
        -1.99549152e-03,  8.82041978e-01],
       [-4.41165279e-02,  3.34882701e-02,  4.97566598e-01,
        -7.50544224e-01,  1.25317878e-02, -3.39154423e-01,
        -3.69594403e-03, -2.66170410e-01],
       [ 1.70064900e-03, -2.30262848e-02, -1.63510759e-01,
        -5.20595149e-01, -2.75798737e-03,  8.31774301e-01,
         6.50749850e-03,  9.90624977e-02],
       [-9.86832609e-01, -4.967885

In [19]:
print(VT.shape)

(8, 8)


In [20]:
S_VT=np.dot(S_diag,VT)

In [21]:
S_VT.shape

(8, 8)

In [22]:
re=S_VT

In [26]:
resize=np.resize(re,(768,768))

In [27]:
resize.shape

(768, 768)

In [30]:
svd_final=np.dot(U,resize)
svd_final

array([[ 2.13920721e+03,  6.97285846e+04,  3.97047147e+04, ...,
         1.50165366e+00, -1.87136829e+02,  3.82714656e-01],
       [ 1.75021267e+03,  5.70491027e+04,  3.24847889e+04, ...,
         1.22859218e+00, -1.53107772e+02,  3.13121625e-01],
       [-3.95707725e+02, -1.28983015e+04, -7.34452569e+03, ...,
        -2.77773911e-01,  3.46163235e+01, -7.07940516e-02],
       ...,
       [-6.95664058e+00, -2.26755360e+02, -1.29118595e+02, ...,
        -4.88333469e-03,  6.08563608e-01, -1.24457710e-03],
       [-3.31339415e+01, -1.08001826e+03, -6.14981889e+02, ...,
        -2.32589457e-02,  2.89854144e+00, -5.92782454e-03],
       [-5.00342368e+00, -1.63089227e+02, -9.28659498e+01, ...,
        -3.51224015e-03,  4.37697123e-01, -8.95137021e-04]])

In [31]:
svd_final.shape

(768, 768)

In [32]:
# splitting the data into testing and training data.

from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(svd_final, y, test_size = 0.2, random_state = 0)

In [33]:
from sklearn.ensemble import RandomForestClassifier
classifier = RandomForestClassifier(max_depth=2, random_state=0)
classifier.fit(X_train, y_train)
y_pred = classifier.predict(X_test) # Predicting test set results

In [34]:
from sklearn.metrics import confusion_matrix
from sklearn.metrics import accuracy_score
cm = confusion_matrix(y_test, y_pred)
print(cm)
print('\nAccuracy: ' + str(accuracy_score(y_test, y_pred)))

[[97 10]
 [35 12]]

Accuracy: 0.7077922077922078


#### Accuracy obtained by using Singular Value Decomposition and Random Forest Classifier is 70%

## PCA

Principal component analysis, or PCA, is a statistical technique to convert high dimensional data to low dimensional data by selecting the most important features that capture maximum information about the dataset. The features are selected on the basis of variance that they cause in the output. 

The feature that causes highest variance is the first principal component. The feature that is responsible for second highest variance isconsidered the second principal component, and so on.

PCA performs best with a normalized feature set. So we have to perform standard scalar normalization to normalize the feature set before applying PCA

In [35]:
# splitting the data into testing and training data.

from sklearn.model_selection import train_test_split

X_train_pca, X_test_pca, y_train_pca, y_test_pca = train_test_split(X, y, test_size = 0.2, random_state = 0)

In [36]:
#Scaling the data
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
X_train_pca = sc.fit_transform(X_train_pca)
X_test_pca = sc.transform(X_test_pca)

In [37]:
#Scaling the data
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
X_train_pca = sc.fit_transform(X_train_pca)
X_test_pca = sc.transform(X_test_pca)

In [38]:
from sklearn.decomposition import PCA # Importing PCA
pca = PCA(n_components=2) #Using 2 principal components
X_train_pca = pca.fit_transform(X_train_pca)
X_test_pca = pca.transform(X_test_pca)

In [39]:
from sklearn.ensemble import RandomForestClassifier
classifier = RandomForestClassifier(max_depth=2, random_state=0)
classifier.fit(X_train_pca, y_train_pca)
y_pred = classifier.predict(X_test_pca) # Predicting test set results


In [40]:
from sklearn.metrics import confusion_matrix
from sklearn.metrics import accuracy_score
cm = confusion_matrix(y_test_pca, y_pred)
print(cm)
print('\nAccuracy: ' + str(accuracy_score(y_test_pca, y_pred)))

[[98  9]
 [29 18]]

Accuracy: 0.7532467532467533


#### With two components of PCA, the random forest algorithm is able to correctly predict 116 out of 154 instances, resulting in 75.3% accuracy

### LDA

Linear discriminant analysis, or LDA, tries to reduce dimensions of the feature set while retaining the information that discriminates output classes. It
is different from PCA in the manner that it relies on the output. It is a supervised technique

In [41]:
from sklearn.discriminant_analysis import LinearDiscriminantAnalysis as LDA
lda = LDA(n_components=1) #Number of components = 1
X_train_l = lda.fit_transform(X_train_pca, y_train_pca)
X_test_l = lda.transform(X_test_pca)
LDA_df = pd.DataFrame(X_test_l)
LDA_df.head()

0
0  1.838308
1 -0.517607
2 -0.940655
3  0.347650
4 -0.911034

In [42]:
from sklearn.ensemble import RandomForestClassifier
classifier = RandomForestClassifier(max_depth=2, random_state=0)
classifier.fit(X_train_l, y_train)
y_pred2 = classifier.predict(X_train_l)
y_pred = classifier.predict(X_test_l)

In [43]:
#Train accuracy
from sklearn.metrics import confusion_matrix
from sklearn.metrics import accuracy_score
cm = confusion_matrix(y_train, y_pred2)
print(cm)
print('\nAccuracy: ' + str(accuracy_score(y_train, y_pred2)))

[[340  53]
 [111 110]]

Accuracy: 0.7328990228013029


In [44]:
#Test accuracy
cm = confusion_matrix(y_test, y_pred)
print(cm)
print('\nAccuracy: ' + str(accuracy_score(y_test, y_pred)))

[[96 11]
 [23 24]]

Accuracy: 0.7792207792207793


#### With one linear discriminant and Random Forest Classfier the model achieved an accuracy of 77.92%.

### The accuracy obtained by using SVD(70%) is less compared to PCA(75%) and LDA(77%). 